File: app/src/main/java/io/intrepid/contest/models/Contest.java
package io.intrepid.contest.models;

import android.os.Parcel;

import com.google.gson.annotations.SerializedName;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.UUID;

import timber.log.Timber;

public class Contest {
    @SerializedName("id")
    private UUID contestId;
    private UUID creatorId;
    @SerializedName("created_at")
    private Date creationDate;
    @SerializedName("updated_at")
    private Date lastUpdatedDate;
    @SerializedName("ended_at")
    private Date endedDate;
    private String title;
    private String description;
    @SerializedName("scoring_categories")
    private List<Category> categories;
    private List<Entry> entries;

    public Contest() {
        categories = new ArrayList<>();
    }

    @Override
    public String toString() {
        return title + "/n" +
                description + "/n" +
                contestId + "/n" +
                creatorId + " " + creationDate +
                " /n size of categories was " + categories.size() +
                "categories " + categories +
                " /n Last updated " + lastUpdatedDate +
                " /n Ended Data " + endedDate;
    }

    public UUID getId() {
        return contestId;
    }

    public void setId(UUID id) {
        this.contestId = id;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public List<Category> getCategories() {
        return categories;
    }

    public void setCategories(List<Category> categories) {
        this.categories = categories;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public List<Entry> getEntries() {
        return entries;
    }

    public void setEntries(List<Entry> entries) {
        this.entries = entries;
    }

    public void setEndedData(Date endedData) {
        this.endedDate = endedData;
    }

    public static class Builder {
        static final int UUID_MIN_LIMIT = 0;
        static final int UUID_MAX_LIMIT = Integer.MAX_VALUE;
        public List<Category> categories = new ArrayList<>();
        public String title;
        public String description;
        UUID contestId;
        UUID creatorId;
        Date creationDate;
        Date lastUpdatedDate;
        Date endedDate;

        public Builder() {
            this(new UUID(UUID_MIN_LIMIT, UUID_MAX_LIMIT));
        }

        Builder(UUID creatorId) {
            this.creatorId = creatorId;
            this.contestId = new UUID(UUID_MIN_LIMIT, UUID_MAX_LIMIT);
        }

        protected Builder(Parcel in) {
            title = in.readString();
            description = in.readString();
        }

        public Builder setTitle(String title) {
            this.title = title;
            return this;
        }

        public Builder setDescription(String description) {
            this.description = description;
            return this;
        }

        public Contest build() {
            if (creationDate == null) {
                creationDate = new Date();
            }
            if (lastUpdatedDate == null) {
                lastUpdatedDate = new Date();
            }
            Timber.d(this.toString());
            Contest contest = new Contest();
            contest.contestId = this.contestId;
            contest.creatorId = this.creatorId;
            contest.lastUpdatedDate = new Date();
            contest.creationDate = this.creationDate;
            contest.endedDate = this.endedDate;
            contest.title = this.title;
            contest.description = this.description;
            contest.categories = this.categories;
            return contest;
        }

        public List<Category> getCategories() {
            return categories;
        }

        public Builder setCategories(List<Category> categories) {
            this.categories = categories;
            return this;
        }
    }
}

File: app/src/main/java/io/intrepid/contest/models/Contest.kt
package io.intrepid.contest.models

import android.os.Parcel
import com.google.gson.annotations.SerializedName
import timber.log.Timber
import java.util.*

open class Contest {
    @SerializedName("id")
    var id: UUID? = null
    var creatorId: UUID? = null
    @SerializedName("created_at")
    var creationDate: Date? = null
    @SerializedName("updated_at")
    var lastUpdatedDate: Date? = null
    @SerializedName("ended_at")
    var endedDate: Date? = null
    set(value) {
        this.endedDate = endedDate
    }
    var title: String? = null
    var description: String? = null
    @SerializedName("scoring_categories")
    var categories: List<Category>? = null
    var entries: List<Entry>? = null
    var submissionsClosedAt: Date? = null
    var participationType: ParticipationType? = null

    init {
        id = UUID.randomUUID()
        categories = ArrayList<Category>()
    }

    override fun toString(): String {
        return title + "/n" +
                description + "/n" +
                id + "/n" +
                creatorId + " " + creationDate +
                " /n size of categories was " + categories!!.size +
                "categories " + categories +
                " /n Last updated " + lastUpdatedDate +
                " /n Ended Data " + endedDate
    }

    class Builder {
        var categories: List<Category> = ArrayList()
        var title: String? = null
        var description: String? = null
        internal var contestId: UUID? = null
        internal var creatorId: UUID? = null
        internal var creationDate: Date? = null
        internal var lastUpdatedDate: Date? = null
        internal var endedDate: Date? = null

        constructor() : this(UUID(UUID_MIN_LIMIT.toLong(), UUID_MAX_LIMIT.toLong())) {}

        internal constructor(creatorId: UUID) {
            this.creatorId = creatorId
            this.contestId = UUID(UUID_MIN_LIMIT.toLong(), UUID_MAX_LIMIT.toLong())
        }

        protected constructor(`in`: Parcel) {
            title = `in`.readString()
            description = `in`.readString()
        }

        fun setTitle(title: String): Builder {
            this.title = title
            return this
        }

        fun setDescription(description: String): Builder {
            this.description = description
            return this
        }

        fun build(): Contest {
            if (creationDate == null) {
                creationDate = Date()
            }
            if (lastUpdatedDate == null) {
                lastUpdatedDate = Date()
            }
            Timber.d(this.toString())
            val contest = Contest()
            contest.id = this.contestId
            contest.creatorId = this.creatorId
            contest.lastUpdatedDate = Date()
            contest.creationDate = this.creationDate
            contest.endedDate = this.endedDate
            contest.title = this.title
            contest.description = this.description
            contest.categories = this.categories
            return contest
        }

        companion object {
            internal val UUID_MIN_LIMIT = 0
            internal val UUID_MAX_LIMIT = Integer.MAX_VALUE
        }
    }
}
